File: stress/stress/backend_toxiproxy.py
"""StressBackend implementation using Toxiproxy for real SP-4 injection."""
from __future__ import annotations

import json
import urllib.request
import urllib.error
from typing import Optional

from stress.stress.backend import (
    StressBackend, NetworkDegradationConfig, ResourcePressureConfig, MemoryStressConfig,
)
from stress.stress.backend_linux import LinuxBackend


class ToxiproxyBackend(StressBackend):
    """Real network fault injection via Toxiproxy REST API.

    Covers SP-4 (network jitter, latency, packet loss) by driving
    Toxiproxy toxics. Delegates other operations to LinuxBackend.

    Requires a running Toxiproxy server and a pre-configured proxy.
    """

    def __init__(
        self, api_url: str = "http://localhost:8474", proxy_name: str = "stress-proxy",
    ) -> None:
        self._api_url = api_url.rstrip("/")
        self._proxy_name = proxy_name
        self._linux = LinuxBackend()
        self._toxic_names: list[str] = []

    def _api(self, method: str, path: str, body: Optional[dict] = None) -> dict:
        url = f"{self._api_url}{path}"
        data = json.dumps(body).encode() if body else None
        req = urllib.request.Request(url, data=data, method=method)
        req.add_header("Content-Type", "application/json")
        try:
            with urllib.request.urlopen(req) as resp:
                return json.loads(resp.read().decode()) if resp.status == 200 else {}
        except urllib.error.HTTPError:
            return {}

    def apply_network_degradation(self, config: NetworkDegradationConfig) -> None:
        self.remove_network_degradation(config.interface)
        # Add latency toxic
        result = self._api("POST", f"/proxies/{self._proxy_name}/toxics", {
            "name": "stress_latency",
            "type": "latency",
            "stream": "downstream",
            "toxicity": 1.0,
            "attributes": {
                "latency": int(config.latency_ms),
                "jitter": int(config.jitter_ms),
            },
        })
        if result:
            self._toxic_names.append("stress_latency")

        # Simulate packet loss via bandwidth limiting + slicer
        if config.loss_percent > 0:
            result = self._api("POST", f"/proxies/{self._proxy_name}/toxics", {
                "name": "stress_timeout",
                "type": "timeout",
                "stream": "downstream",
                "toxicity": config.loss_percent / 100.0,
                "attributes": {"timeout": 1},
            })
            if result:
                self._toxic_names.append("stress_timeout")

    def remove_network_degradation(self, interface: str) -> None:
        for name in self._toxic_names:
            self._api("DELETE", f"/proxies/{self._proxy_name}/toxics/{name}")
        self._toxic_names.clear()

    def apply_resource_pressure(self, config: ResourcePressureConfig) -> None:
        self._linux.apply_resource_pressure(config)

    def remove_resource_pressure(self, cgroup_path: str) -> None:
        self._linux.remove_resource_pressure(cgroup_path)

    def inject_memory_stress(self, config: MemoryStressConfig) -> None:
        raise NotImplementedError("ToxiproxyBackend only handles network faults. Use StressNgBackend for memory stress.")

    def remove_memory_stress(self) -> None:
        pass

    def pause_workload(self, pid: int) -> None:
        self._linux.pause_workload(pid)

    def resume_workload(self, pid: int) -> None:
        self._linux.resume_workload(pid)

    def apply_network_partition(self, interface: str) -> None:
        # Use Toxiproxy: disable the proxy entirely
        self._api("POST", f"/proxies/{self._proxy_name}", {"enabled": False})

    def remove_network_partition(self, interface: str) -> None:
        self._api("POST", f"/proxies/{self._proxy_name}", {"enabled": True})

    def name(self) -> str:
        return "toxiproxy"

File: stress-ref/src/stress/backend_toxiproxy.rs
#[cfg(feature = "toxiproxy")]
use crate::stress::backend::*;

#[cfg(feature = "toxiproxy")]
pub struct ToxiproxyBackend {
    api_url: String,
    proxy_name: String,
    toxic_names: Vec<String>,
}

#[cfg(feature = "toxiproxy")]
impl ToxiproxyBackend {
    pub fn new(api_url: &str, proxy_name: &str) -> Self {
        Self {
            api_url: api_url.trim_end_matches('/').to_string(),
            proxy_name: proxy_name.to_string(),
            toxic_names: Vec::new(),
        }
    }

    fn api_post(&self, path: &str, body: &serde_json::Value) -> Result<serde_json::Value, String> {
        ureq::post(&format!("{}{}", self.api_url, path))
            .set("Content-Type", "application/json")
            .send_json(body.clone())
            .map_err(|e| format!("Toxiproxy API error: {e}"))
            .and_then(|resp| {
                resp.into_json::<serde_json::Value>()
                    .map_err(|e| format!("JSON parse error: {e}"))
            })
    }

    fn api_delete(&self, path: &str) -> Result<(), String> {
        ureq::delete(&format!("{}{}", self.api_url, path))
            .call()
            .map_err(|e| format!("Toxiproxy API error: {e}"))?;
        Ok(())
    }
}

#[cfg(feature = "toxiproxy")]
impl StressBackend for ToxiproxyBackend {
    fn apply_network_degradation(&mut self, config: &NetworkDegradationConfig) -> Result<(), String> {
        self.remove_network_degradation(&config.interface)?;
        let body = serde_json::json!({
            "name": "stress_latency",
            "type": "latency",
            "stream": "downstream",
            "toxicity": 1.0,
            "attributes": {
                "latency": config.latency_ms as i64,
                "jitter": config.jitter_ms as i64,
            }
        });
        self.api_post(&format!("/proxies/{}/toxics", self.proxy_name), &body)?;
        self.toxic_names.push("stress_latency".to_string());
        Ok(())
    }

    fn remove_network_degradation(&mut self, _interface: &str) -> Result<(), String> {
        for name in self.toxic_names.drain(..) {
            let _ = self.api_delete(&format!("/proxies/{}/toxics/{}", self.proxy_name, name));
        }
        Ok(())
    }

    fn apply_resource_pressure(&mut self, _config: &ResourcePressureConfig) -> Result<(), String> {
        Err("ToxiproxyBackend only handles network faults".into())
    }

    fn remove_resource_pressure(&mut self, _cgroup_path: &str) -> Result<(), String> {
        Ok(())
    }

    fn inject_memory_stress(&mut self, _config: &MemoryStressConfig) -> Result<(), String> {
        Err("ToxiproxyBackend only handles network faults".into())
    }

    fn remove_memory_stress(&mut self) -> Result<(), String> { Ok(()) }

    fn pause_workload(&mut self, _pid: u32) -> Result<(), String> {
        Err("ToxiproxyBackend only handles network faults".into())
    }

    fn resume_workload(&mut self, _pid: u32) -> Result<(), String> { Ok(()) }

    fn apply_network_partition(&mut self, _interface: &str) -> Result<(), String> {
        self.api_post(
            &format!("/proxies/{}", self.proxy_name),
            &serde_json::json!({"enabled": false}),
        )?;
        Ok(())
    }

    fn remove_network_partition(&mut self, _interface: &str) -> Result<(), String> {
        self.api_post(
            &format!("/proxies/{}", self.proxy_name),
            &serde_json::json!({"enabled": true}),
        )?;
        Ok(())
    }

    fn name(&self) -> &str { "toxiproxy" }
}
